packages/botonic-core/src/core-bot.ts

```diff
@@ -91,194 +91,160 @@ export class CoreBot {
+      params: {},
+      lastRoutePath,
+      plugins: this.plugins,
+      defaultTyping: this.defaultTyping,
+      defaultDelay: this.defaultDelay,
+      input,
+      session,
+    }
+
+    const output = await this.runInput(requestCoreContext)
 
     if (session._botonic_action?.startsWith(BotonicAction.Redirect)) {
-      return await this.runRedirectAction(output, {
-        input,
-        session,
-        lastRoutePath,
-      })
+      return await this.runRedirectAction(output, requestCoreContext)
     }
 
     return output
   }
 
-  private async runInput({
-    input,
-    session,
-    lastRoutePath,
-  }: BotRequest): Promise<BotResponse> {
-    session = session || {}
-    if (!session.__locale) session.__locale = 'en'
+  private async runInput(
+    requestCoreContext: RequestCoreContext
+  ): Promise<BotResponse> {
+    requestCoreContext.session = requestCoreContext.session || {}
+    if (!requestCoreContext.session.__locale)
+      requestCoreContext.session.__locale = 'en'
 
-    if (input.type === INPUT.CHAT_EVENT) {
+    if (requestCoreContext.input.type === INPUT.CHAT_EVENT) {
       return {
-        input,
-        session,
-        lastRoutePath,
+        input: requestCoreContext.input,
+        session: requestCoreContext.session,
+        lastRoutePath: requestCoreContext.lastRoutePath,
         response: [],
       }
     }
 
-    await this.runPrePlugins(input, session, lastRoutePath)
-
-    const output = await this.getOutput(input, session, lastRoutePath)
+    await this.runPrePlugins(requestCoreContext)
 
+    const output = await this.getOutput(requestCoreContext)
+    requestCoreContext = this.updateRequestFromOutput(
+      requestCoreContext,
+      output
+    )
     const response = await this.renderer({
-      request: this.createRequestFromOutput(
-        { input, session, lastRoutePath },
-        output
-      ),
+      request: requestCoreContext,
       actions: [output.fallbackAction, output.action, output.emptyAction],
     })
 
-    lastRoutePath = output.lastRoutePath
-
-    await this.runPostPlugins(input, session, lastRoutePath, response)
+    await this.runPostPlugins(requestCoreContext, response)
 
-    session.is_first_interaction = false
+    requestCoreContext.session.is_first_interaction = false
 
     return {
-      input,
+      input: requestCoreContext.input,
       response,
-      session,
-      lastRoutePath,
+      session: requestCoreContext.session,
+      lastRoutePath: requestCoreContext.lastRoutePath,
     }
   }
 
-  private async runPrePlugins(
-    input: Input,
-    session: Session,
-    lastRoutePath: RoutePath
-  ) {
+  private async runPrePlugins(requestCoreContext: RequestCoreContext) {
     if (this.plugins) {
-      await runPlugins(
-        this.plugins,
-        'pre',
-        input,
-        session,
-        lastRoutePath,
-        undefined
-      )
+      await runPlugins({ requestCoreContext, mode: 'pre' })
```

Comment:
we could have 'pre' and 'post' typed as `type PluginMode = 'pre' | 'post'`

Reply:
Yes but this is a type not an enum, if you try to pass a string that is neither `pre` nor `post` there is a type error. But you can not do PluginMode.pre